### PR TITLE
[InstCombine] Fix buggy transform in foldNestedSelects; PR 71330

### PR DESCRIPTION
**Pull request by goldsteinn**
- [InstCombine] Add reduced test for PR 71330; NFC
- [InstCombine] Fix buggy transform in `foldNestedSelects`; PR 71330

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/pr71330.ll [hunk 1]
```diff
@@ -0,0 +1,229 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 3
+; RUN: opt -O3 -S < %s | FileCheck %s
```
Comment (goldsteinn) [inline llvm/test/Transforms/InstCombine/pr71330.ll:2]: I know this isn't ideal.
I'm properly not able to get a reproduction without `-O3`.
Even if I enable all O3 passes manually it doesn't reproduce (neither the original bug nor this).
I'm thinking that may mean I misdiagnosed something here.

Comment (nikic) [reply llvm/test/Transforms/InstCombine/pr71330.ll:2]: Have you tried `-preserve-ll-uselistorder`?

Comment (nikic) [reply llvm/test/Transforms/InstCombine/pr71330.ll:2]: How did you determine `<all-O3-passes>`?

Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/pr71330.ll:2]:
```
./bin/opt -O3 --print-pipeline-passes -
```
Which results in:
```
annotation2metadata,forceattrs,inferattrs,coro-early,function<eager-inv>(lower-expect,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;no-switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-sink-common-insts;speculate-blocks;simplify-cond-branch>,sroa<modify-cfg>,early-cse<>,callsite-splitting),openmp-opt,ipsccp,called-value-propagation,globalopt,function<eager-inv>(mem2reg,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-sink-common-insts;speculate-blocks;simplify-cond-branch>),always-inline,require<globals-aa>,function(invalidate<aa>),require<profile-summary>,cgscc(devirt<4>(inline,function-attrs<skip-non-recursive-function-attrs>,argpromotion,openmp-opt-cgscc,function<eager-inv;no-rerun>(sroa<modify-cfg>,early-cse<memssa>,speculative-execution,jump-threading,correlated-propagation,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-sink-common-insts;speculate-blocks;simplify-cond-branch>,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>,aggressive-instcombine,libcalls-shrinkwrap,tailcallelim,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-sink-common-insts;speculate-blocks;simplify-cond-branch>,reassociate,constraint-elimination,loop-mssa(loop-instsimplify,loop-simplifycfg,licm<no-allowspeculation>,loop-rotate<header-duplication;no-prepare-for-lto>,licm<allowspeculation>,simple-loop-unswitch<nontrivial;trivial>),simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-sink-common-insts;speculate-blocks;simplify-cond-branch>,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>,loop(loop-idiom,indvars,loop-deletion,loop-unroll-full),sroa<modify-cfg>,vector-combine,mldst-motion<no-split-footer-bb>,gvn<>,sccp,bdce,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>,jump-threading,correlated-propagation,adce,memcpyopt,dse,move-auto-init,loop-mssa(licm<allowspeculation>),coro-elide,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;hoist-common-insts;sink-common-insts;speculate-blocks;simplify-cond-branch>,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>),function-attrs,function(require<should-not-run-function-passes>),coro-split)),deadargelim,coro-cleanup,globalopt,globaldce,elim-avail-extern,rpo-function-attrs,recompute-globalsaa,function<eager-inv>(float2int,lower-constant-intrinsics,chr,loop(loop-rotate<header-duplication;no-prepare-for-lto>,loop-deletion),loop-distribute,inject-tli-mappings,loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>,infer-alignment,loop-load-elim,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>,simplifycfg<bonus-inst-threshold=1;forward-switch-cond;switch-range-to-icmp;switch-to-lookup;no-keep-loops;hoist-common-insts;sink-common-insts;speculate-blocks;simplify-cond-branch>,slp-vectorizer,vector-combine,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>,loop-unroll<O3>,transform-warning,sroa<preserve-cfg>,infer-alignment,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>,loop-mssa(licm<allowspeculation>),alignment-from-assumptions,loop-sink,instsimplify,div-rem-pairs,tailcallelim,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-sink-common-insts;speculate-blocks;simplify-cond-branch>),globaldce,constmerge,cg-profile,rel-lookup-table-converter,function(annotation-remarks),verify
```

Comment (nikic) [reply llvm/test/Transforms/InstCombine/pr71330.ll:2]: Fixed in https://github.com/llvm/llvm-project/commit/6e56c35d1959295289734baf8924d477f770d6f8.

Comment (nikic) [reply llvm/test/Transforms/InstCombine/pr71330.ll:2]: I believe this is an (unreduced) InstCombine-only test case: https://gist.github.com/nikic/a7f94a7549896284b8444bea9b4a5780

Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/pr71330.ll:2]: Thanks :)

Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/pr71330.ll:2]: Hmm, doesn't really reduce.

Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/pr71330.ll:2]: It does by hand, `llvm-reduce` doesn't do well pairing down use-lists.

File: llvm/test/Transforms/InstCombine/pr71330.ll [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+  %cmp = icmp slt i32 %conv35, 0
+  br i1 %cmp, label %for.cond7.preheader, label %for.end36
+
+for.end36:                                        ; preds = %for.inc32
+  %storemerge.lcssa = phi i32 [ %conv35, %for.inc32 ]
+  ret void
+
+; uselistorder directives
+  uselistorder i32 %storemerge33, { 2, 4, 1, 3, 0 }
+}
```
Comment (goldsteinn) [inline llvm/test/Transforms/InstCombine/pr71330.ll:230]: I ended up reducing by hand, `llvm-reduce` simply doesn't change anything.

Comment (nikic) [reply llvm/test/Transforms/InstCombine/pr71330.ll:230]: What kind of interestingness test were you using? Diff between two opt binaries?

Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/pr71330.ll:230]: I put an assertion in at the point of the bug and any change that triggered the assert.

Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/pr71330.ll:230]:
```
#! /usr/bin/env python3

import os
import sys
from common_util import run_proc


def os_do(cmd):
    print(cmd)
    return os.system(cmd)


s, r, stdout, stderr = run_proc(
    "./bin/opt -passes=instcombine {} -o unused.ll".format(sys.argv[1]),
    timeout=10,
    decode=True)
print("R={}".format(r))
if s is False:
    sys.exit(1)
if r != 134:
    sys.exit(1)
print(stderr)
if "UID:1234567890" in stderr:
    sys.exit(0)

sys.exit(1)

```

Comment (nikic) [reply llvm/test/Transforms/InstCombine/pr71330.ll:230]: Hm yeah, for some reason llvm-reduce is completely ineffective for this case. What worked for me is to manually remove uses of storemerge33 and then reorder them so the uselistorder is no longer necessary, and then run llvm-reduce afterwards. This gave me: https://gist.github.com/nikic/da8251d05a5df6038bd3e2eb773f64c0 This is probably still nowhere near minimal, but maybe good enough for a test case.

Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/pr71330.ll:230]: Updated to your test case.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by goldsteinn** [comment]
Think this needs to be backported to 17 release. Im checking if the repro still applies. Otherwise will req cherry-pick w.o the test case.

**Pull request closed without merging**